File: exercises/exercise-1-testing-java/src/main/java/io/miragon/example/domain/NewsletterSubscription.java
package io.miragon.example.domain;

import java.time.LocalDateTime;
import java.util.UUID;

public record NewsletterSubscription(
        SubscriptionId id,
        Name name,
        Email email,
        NewsletterId newsletter,
        LocalDateTime registrationDate,
        SubscriptionStatus status
) {
    public NewsletterSubscription(Name name, Email email, NewsletterId newsletter) {
        this(
                new SubscriptionId(UUID.randomUUID()),
                name,
                email,
                newsletter,
                LocalDateTime.now(),
                SubscriptionStatus.PENDING
        );
    }

    public NewsletterSubscription confirmRegistration() {
        return new NewsletterSubscription(
                id,
                name,
                email,
                newsletter,
                registrationDate,
                SubscriptionStatus.CONFIRMED
        );
    }

    public NewsletterSubscription abortRegistration() {
        return new NewsletterSubscription(
                id,
                name,
                email,
                newsletter,
                registrationDate,
                SubscriptionStatus.ABORTED
        );
    }
}

File: exercises/exercise-1-testing-kotlin/src/main/kotlin/de/emaarco/example/domain/newsletter/NewsletterSubscription.kt
package de.emaarco.example.domain.newsletter

import de.emaarco.example.domain.Email
import de.emaarco.example.domain.Name
import de.emaarco.example.domain.SubscriptionId
import de.emaarco.example.domain.SubscriptionStatus
import java.time.LocalDateTime
import java.util.*

data class NewsletterSubscription(
    val id: SubscriptionId = SubscriptionId(UUID.randomUUID()),
    val name: Name,
    val email: Email,
    val newsletter: NewsletterId,
    val registrationDate: LocalDateTime = LocalDateTime.now(),
    val status: SubscriptionStatus = SubscriptionStatus.PENDING,
) {
    fun confirmRegistration() = this.copy(status = SubscriptionStatus.CONFIRMED)
    fun abortRegistration() = this.copy(status = SubscriptionStatus.ABORTED)
}
